Design/Workbench/design-development.md

```diff
@@ -0,0 +1,22 @@
+# Development
+
+## Coding Style
+
+It is proposed we introduce a new Python style guide for Mantid. All new and refactored code should then follow this guide. The [IPython style guide][ipython-style] offers quite nice guidelines and it is suggested that we
+follow these guidelines where applicable, including a template file for new code that could be used in the `class_maker` to generate new files.
+
+## Static Analysis
+
+We will no longer be able to rely on a compiler to catch simply errors - it is therefore imperative that we use static analysis to its fullest extent. While `flake8` is good it is not as thorough as `pylint` and given our
```

Comment:
My main concern is we will now be developing a fully fledged application in Python and will not have a compiler to catch basic mistakes. `flake8` is okay but `pylint` does a far better job at genuine error detection.

Consider this:
```
def get_volumes():
    return [1, 2, 3, 4, 5]

def get_other_volumes():
    return [1, 2, 3, 4, 5, 6]

def do_something(volume):
    pass

def do_something_else(volume):
    pass

for volume in get_volumes():
    do_something(volume)

for volme in get_other_volumes():
    do_something_else(volume)
```

It contains a subtle bug that `pylint` detects out of the box yet `flake8` gives it a clean bill of health. Pylint's error:

```
W: 22,22: Using possibly undefined loop variable 'volume' (undefined-loop-variable)
```

The bug is that there is a typo in the the second loop variable so the loop actually uses `volume` from the previous loop definition as python 2 has terrible scope rules. A C++ compiler would not even compile this code.

One could argue that a unit test should pick up this type of thing but I think pylint's warning is much easier to interpret and find where the problem lies.

On the speed side of things I was thinking that we could do something similar to the doctests/system tests in that we run it combined with one of the builds so that we don't have to build separately - this is much of what caused the slowdown before.

@mantidproject/gatekeepers Any thoughts here?